### ISSUE TITLE
Unable to log into Twitch with Ruffle enabled

### ISSUE THREAD

**Issue by Azure-Fang** [ruffle-rs/ruffle#23929]
### Describe the bug

With Ruffle enabled, attempting to log into https://twitch.tv throws an "unsupported browser" error. With Ruffle (or just its permissions) disabled, login works as normal.

### Expected behavior

Twitch.tv supports the two most recent versions of Firefox for login. As long as the browser version is detected, login is possible.

### Content Location

Affecting non-Flash content: https://twitch.tv

### Affected platform

Browser's extension

### Operating system

Windows 10

### Browser

Firefox 151.0.3 (64-bit)

### Additional information

**Reproduction:**
1. Log out of https://twitch.tv and delete site cache (this will clear 30-day authentication cookie).
2. Install and enable Ruffle extension.
3. Restart browser.
4. Attempt to log in on https://twitch.tv. Login will read "Your browser is not currently supported." and will fail.
5. Disable Ruffle extension.
6. Reload Twitch tab.
7. Attempt to login. Login will be successful.

This issue began recently. I haven't kept track of Ruffle's version releases, but it started happening some time on or before May 29th.

**Comment by nroach44** [comment]
I can confirm this, I'm able to log in via a Private session, and when the Ruffle plugin is disabled in firefox.

**Comment by torokati44** [comment]
Hmmm, could it be https://github.com/ruffle-rs/ruffle/pull/22449 ?

**Comment by KandiPanda** [comment]
Also, when already logged in: 
 - prevents claiming bonus points (the +50 ones) 
 - prevents notifications from appearing in the notifications dropdown 
 - causes Drops & Rewards > All Campaigns to claim that there are no current campaigns (when there are actually many) 
 - prevents much of the recently-added Drops interface in the chat area from working (though drops can still be earned, and then claimed in Drops inventory)

Started doing this maybe a week or so ago, or maybe a little earlier than that?

Browser: 
Firefox 115.36.0esr (Yes, that's still an actively updated branch, amazingly enough!) 

(Side thought: Would adding a site blacklist that users could specify sites to disable Ruffle on (mainly sites that misbehave with Ruffle present, especially ones with no Flash content anyways) be a potential option?)

**Comment by SilentSpud** [comment]
> Hmmm, could it be [#22449](https://github.com/ruffle-rs/ruffle/pull/22449) ?

I converted that commit into a userscript that polyfills `document.embeds` and I was still able to log in just fine

<details>

<summary>Userscript polyfill-document-embeds.user.js</summary>

```javascript
// ==UserScript==
// @name        Ruffle polyfill test
// @namespace   Violentmonkey Scripts
// @version     1.0.0
// @match       https://www.twitch.tv/*
// @grant       none
// @author      Silent Spud
// @run-at      document-start
// ==/UserScript==
// polyfill-document-embeds.ts
function mergeSorted(a, b, comparator) {
  const result = [];
  let i = 0;
  let j = 0;
  while (i < a.length && j < b.length) {
    const x = a[i];
    const y = b[j];
    if (comparator(x, y) <= 0) {
      result.push(x);
      i++;
    } else {
      result.push(y);
      j++;
    }
  }
  while (i < a.length) {
    result.push(a[i++]);
  }
  while (j < b.length) {
    result.push(b[j++]);
  }
  return result;
}
function domComesBefore(a, b) {
  if (a === b) {
    return 0;
  }
  const pos = a.compareDocumentPosition(b);
  if (pos & Node.DOCUMENT_POSITION_FOLLOWING) {
    return -1;
  }
  if (pos & Node.DOCUMENT_POSITION_PRECEDING) {
    return 1;
  }
  return 0;
}
function getSelector(tries) {
  const selectors = ["ruffle-embed"];
  for (let i = 1;i <= tries; i++) {
    selectors.push(`ruffle-embed-${i}`);
  }
  return selectors.join(", ");
}
function polyfillDocumentEmbeds(tries) {
  const orig = Object.getOwnPropertyDescriptor(Document.prototype, "embeds");
  if (!orig?.get) {
    return;
  }
  const CACHE_SYM = Symbol("ruffle_embeds_cache");
  Object.defineProperty(Document.prototype, "embeds", {
    get() {
      const documentWithCache = this;
      const existing = documentWithCache[CACHE_SYM];
      if (existing) {
        return existing;
      }
      let cachedNodes = null;
      const getFreshNodes = () => {
        const baseEmbeds = orig.get.call(this);
        const selector = getSelector(tries);
        const ruffleEmbeds = Array.from(this.querySelectorAll(selector));
        return mergeSorted(Array.from(baseEmbeds), ruffleEmbeds, domComesBefore);
      };
      const nodes = () => {
        if (cachedNodes !== null) {
          return cachedNodes;
        }
        cachedNodes = getFreshNodes();
        queueMicrotask(() => {
          cachedNodes = null;
        });
        return cachedNodes;
      };
      const base = Object.create(HTMLCollection.prototype);
      Object.defineProperty(base, "length", {
        enumerable: true,
        configurable: true,
        get() {
          return nodes().length;
        }
      });
      base.item = function(index) {
        return nodes()[index] ?? null;
      };
      base.namedItem = function(name) {
        const list = nodes();
        for (const el of list) {
          const htmlEl = el;
          if (name && (htmlEl.getAttribute("name") === name || htmlEl.id === name)) {
            return htmlEl;
          }
        }
        return null;
      };
      base[Symbol.iterator] = function* () {
        for (const el of nodes()) {
          yield el;
        }
      };
      const proxy = new Proxy(base, {
        get(target, prop, receiver) {
          if (typeof prop === "string") {
            const index = Number(prop);
            if (!Number.isNaN(index) && index >= 0) {
              return nodes()[index];
            }
          }
          return Reflect.get(target, prop, receiver);
        },
        has(target, prop) {
          if (typeof prop === "string") {
            const index = Number(prop);
            if (!Number.isNaN(index) && index >= 0) {
              return index < nodes().length;
            }
          }
          return Reflect.has(target, prop);
        },
        ownKeys() {
          const len = nodes().length;
          const keys = [];
          for (let i = 0;i < len; i++) {
            keys.push(String(i));
          }
          return keys;
        },
        getOwnPropertyDescriptor(target, prop) {
          if (typeof prop === "string") {
            const index = Number(prop);
            if (!Number.isNaN(index) && index >= 0 && index < nodes().length) {
              return {
                enumerable: true,
                configurable: true,
                writable: false,
                value: nodes()[index]
              };
            }
          }
          return Reflect.getOwnPropertyDescriptor(target, prop);
        }
      });
      proxy[CACHE_SYM] = true;
      documentWithCache[CACHE_SYM] = proxy;
      return proxy;
    },
    configurable: true,
    enumerable: true
  });
}
polyfillDocumentEmbeds();
```
</details>

**Comment by emmanu888** [comment]
Ugh. Had the same issue on Chrome. Ruffle was the cause, something about Ruffle is causing Twitch to go all out of whack.

**Comment by cj-wong** [comment]
This issue specifically affects Twitch's GraphQL API. I'm not a JS developer so I can't pinpoint further than that.

**Comment by NWPlayer123** [comment]
> Also, when already logged in:
> 
>     * prevents claiming bonus points (the +50 ones)
> 
>     * prevents notifications from appearing in the notifications dropdown
> 
>     * causes Drops & Rewards > All Campaigns to claim that there are no current campaigns (when there are actually many)
> 
>     * prevents much of the recently-added Drops interface in the chat area from working (though drops can still be earned, and then claimed in Drops inventory)
> 
> 
> Started doing this maybe a week or so ago, or maybe a little earlier than that?
> 
> Browser: Firefox 115.36.0esr (Yes, that's still an actively updated branch, amazingly enough!)
> 
> (Side thought: Would adding a site blacklist that users could specify sites to disable Ruffle on (mainly sites that misbehave with Ruffle present, especially ones with no Flash content anyways) be a potential option?)

Also causes creator pages like Roles to not load, +1 on all those symptoms as well, I'm on 151.0.4 on CachyOS

**Comment by laymonage** [comment]
This also causes the viewers list to not load. The extension causes the browser to fail Twitch's integrity check, which happens as part of GraphQL requests for the aforementioned features.

**Comment by torokati44** [comment]
Strange, *.twitch.tv is already supposed to be blocklisted...?
https://github.com/ruffle-rs/ruffle/pull/8150

**Comment by TsukiZero** [comment]
It breaks nearly EVERYTHING Twitch: Redeems, Login, Drops, etc.

**Comment by SirJorb** [comment]
Adding another confirmation of the same. While logged in from an existing session, I couldn't even follow someone.

**Comment by TsukiZero** [comment]
> Adding another confirmation of the same. While logged in from an existing session, I couldn't even follow someone.

Adding to that, you also can't unfollow.

**Comment by ElTipejoLoco** [comment]
I don't think detailing fully all that Twitch's GraphQL integrity failure affects (i.e.- "notifications and 'stories' also never finish loading") would help the Ruffle devs any more than them figuring out why their apparently built-in 'don't run on Twitch' code is apparently not working.

Maybe the `*.twitchcdn.net/*` URLs also need to be matched?

**Comment by kjarosh** [comment]
`*.twitchcdn.net` indeed appears to be the culprit. Hopefully https://github.com/ruffle-rs/ruffle/pull/24114 will fix it. Tested logging in and claiming bonus points.
